pkg/e2etests/metrics_e2e_test.go

```diff
@@ -0,0 +1,1186 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package e2etests
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/siglens/siglens/pkg/common/dtypeutils"
+	"github.com/siglens/siglens/pkg/config"
+	"github.com/siglens/siglens/pkg/integrations/prometheus/promql"
+	"github.com/siglens/siglens/pkg/segment"
+	"github.com/siglens/siglens/pkg/segment/memory/limit"
+	"github.com/siglens/siglens/pkg/segment/query"
+	"github.com/siglens/siglens/pkg/segment/query/metadata"
+	"github.com/siglens/siglens/pkg/segment/results/mresults"
+	"github.com/siglens/siglens/pkg/segment/utils"
+	"github.com/siglens/siglens/pkg/segment/writer"
+	"github.com/siglens/siglens/pkg/segment/writer/metrics"
+	"github.com/siglens/siglens/pkg/segment/writer/metrics/meta"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+/*
+	This test suite is to test the metrics query E2E.
+
+	# Whenever a new test is added, do not forget to call the cleanUp function.
+	All the tests below will have `defer CleanUp(t)` as the first line at the start of any testing function.
+
+	Steps to follow for adding a new metric query test:
+	1. Define the Test function name appropriately.
+	2. defer CleanUp(t) as the first line of the test function. This will ensure that the data is cleaned up after the test is run.
+	3. init the testing config by calling initTestConfig()
+	4. Get the test metrics data by calling GetTestMetricsData(startTimestamp). This function returns the test metrics timeseries data, metric names, tag keys and tag key values.
+	5. Ingest the test metrics data by calling ingestTestMetricsData(allTimeSeries). This function ingests the test metrics data into the in-memory buffer.
+	6. Rotate the metrics data by calling rotateMetricsDataAndClearSegStore(true). This function forces the rotation of the metrics data. And clear the unrotated segments in memory.
+	7. Initialize the metrics data and query node by calling initializeMetricsMetaData(). This function populates the metrics data.
+	8. Define the time range for the query by creating a MetricsTimeRange object.
+	9. Define the expected results for the query.
+	10. Parse the query by calling ConvertPromQLToMetricsQuery. This function converts the promql query to metrics query.
+	11. Execute the metrics query by calling ExecuteMetricsQuery/ExecuteMultipleMetricsQuery depening on the query Type. This function executes the metrics query and returns the results.
+	12. Validate the results by comparing the expected results with the actual results.
+*/
+
+type timeSeries struct {
+	Metric    string            `json:"metric"`
+	Tags      map[string]string `json:"tags"`
+	Timestamp uint32            `json:"timestamp"`
+	Value     int               `json:"value"`
+}
+
+const dataStartTimestamp uint32 = 1718052279
```

Comment:
Is this arbitrary, or should it be a recent unix epoch? If it is arbitrary I'd recommend saying so in a comment and optionally setting it to a different value (e.g., 42, or 1 billion) to make it more obvious

Reply:
Yes, it is kind of arbitrary. At first, I kept it dynamic, setting it to `24h` before, but that sometimes is failing making the test to be flaky. 
You might have seen some comments where I have that at T1 and T2, the time diff is less than 5 seconds, so they will aggregated into the same bucket. And sometimes that downsampling is not happening, when I set it to dynamic.

As our downsampling formula = `downsampledTime = (ts / s.dsSeconds) * s.dsSeconds` , So sometimes they are not ending up under the same bucket and causing my test to be failed. 

So then I changed it to be static. 
And actually this epoch timestamp does not matter, it is just so that the data generation will use this as start timestamp and the queries will also use this as start timestamp. So as long as both of these use the same timestamp it should be fine, given that T1 and T2 will end up in the same bucket. 

I will write some comment, explaining this.

Reply:
Ok, and where is the downsampling interval getting set? is it by `ConvertPromQLToMetricsQuery()`? What does it get set to for your tests? You want `dataStartTimestamp` to be set so that items at `+1` and `+2` from that get put into the same bucket, right? This info might be useful in the comment you've added

Reply:
Yes, it is being set in `ConvertPromQLToMetricsQuery`. And the time range I am setting when making the query is 1 hour. As the max data ingested is at `startTimestamp + 3600`. 

And Yes, I have added this information in the comment above `dataStartTimestamp` variable.